From b5013893fc8fb4749788c6fdba4bd6ee91f2ce2b Mon Sep 17 00:00:00 2001
From: Zheng Yejian <zhengyejian1@huawei.com>
Date: Thu, 27 Oct 2022 18:36:48 +0800
Subject: [PATCH] livepatch/core: Fix livepatch/state leak on error path
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: b5013893fc8fb4749788c6fdba4bd6ee91f2ce2b
Modified-by-SEL: No


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I5WF0G
CVE: NA

--------------------------------

File '/proc/livepatch/state' should be removed before removing
'/proc/livepatch', otherwise following log will appear:
  remove_proc_entry: removing non-empty directory '/proc/livepatch',
  leaking at least 'state'

Fixes: c33e42836a74 ("livepatch/core: Allow implementation without ftrace")
Signed-off-by: Zheng Yejian <zhengyejian1@huawei.com>
Reviewed-by: Kuohai Xu <xukuohai@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 kernel/livepatch/core.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/kernel/livepatch/core.c b/kernel/livepatch/core.c
index 780a825cee8b..f6981faa18a8 100644
--- a/kernel/livepatch/core.c
+++ b/kernel/livepatch/core.c
@@ -2331,13 +2331,15 @@ static int __init klp_init(void)
 
 	klp_root_kobj = kobject_create_and_add("livepatch", kernel_kobj);
 	if (!klp_root_kobj)
-		goto error_remove;
+		goto error_remove_state;
 
 #ifdef CONFIG_LIVEPATCH_STOP_MACHINE_CONSISTENCY
 	arch_klp_init();
 #endif
 	return 0;
 
+error_remove_state:
+	remove_proc_entry("livepatch/state", NULL);
 error_remove:
 	remove_proc_entry("livepatch", NULL);
 error_out:
-- 
2.33.0

